### PR TITLE
[tune] Add optional Experiment to Searcher/SearchAlgo.

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

Add optional Experiment to Searcher/SearchAlgo for them to leverage experiment settings.

## Related issue number

#17346 

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by xwjiang2010** [comment]
Before I forget, what is the difference between Searcher and SearchAlgorithm? I am sure Searcher is public facing, how about SearchAlgorithm?

**Review by Yard1 — CHANGES_REQUESTED** [review]
Thanks! I have a few comments, let me know what you think

File: python/ray/tune/suggest/ax.py [hunk 1]
```diff
@@ -240,6 +244,7 @@ def set_search_properties(self, metric: Optional[str], mode: Optional[str],
             self._metric = metric
         if mode:
             self._mode = mode
+        self._experiments = experiments if experiments else []
```
[the remark this answers is not in the document]
Comment (xwjiang2010) [reply python/ray/tune/suggest/ax.py:247]: Sounds good. I will leave it to custom searchers' own impl.

File: python/ray/tune/suggest/util.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+from ray.tune.experiment import Experiment
+
+logger = logging.getLogger(__name__)
+
+
+def set_search_properties_backwards_compatible(
+        set_search_properties_func, metric: Optional[str], mode: Optional[str],
+        config: Dict, experiment: Optional[Experiment]) -> bool:
+    try:
```
Comment (Yard1) [inline python/ray/tune/suggest/util.py:12]: Can we add a short docstring explaining what this is for? One, two sentences

Comment (xwjiang2010) [reply python/ray/tune/suggest/util.py:16]: Added :)

**Review by Yard1 — APPROVED** [review]
Is it possible for the experiment to be None? If not, we should remove the Optional type hint. Great work, thanks!

**Comment by xwjiang2010** [comment]
> Is it possible for the experiment to be None? If not, we should remove the Optional type hint. Great work, thanks!

Originally I thought we (Ray Tune) have full control of the caller side. However, apparently flaml is also calling set_search_properties into Optuna Searcherer (https://microsoft.github.io/FLAML/_modules/flaml/searcher/blendsearch.html). So I find myself in a weird situation where Tune will use BlendSearcher, which in turn, uses OptunaSearcher under the hood. set_search_properties is also propagating through from Tune to BlendSearcher[1] to OptunaSearcher[2]. [1] is fine with try catch, [2] is not as it is MSR's own code. So I guess I have to leave this field optional for now.

**Comment by Yard1** [comment]
That's fine by me, thanks!

File: python/ray/tune/suggest/suggestion.py [hunk 3]
```diff
@@ -123,11 +128,14 @@ def set_search_properties(self, metric: Optional[str], mode: Optional[str],
         Args:
             metric (str): Metric to optimize
             mode (str): One of ["min", "max"]. Direction to optimize.
-            config (dict): Tune config dict.
+            config (dict): Tune config dict. This should ideally be used for
+            passing algorithm-specific parameters.
+            experiment (Optional[Experiment]): This is used for passing
+            algorithm-agnostic parameters. User should not need to touch this.
```
Comment (richardliaw) [inline python/ray/tune/suggest/suggestion.py:134]: The overflow lines need to be indented:
```suggestion
            config (dict): Tune config dict. This should ideally be used for
                passing algorithm-specific parameters.
            experiment (Optional[Experiment]): This is used for passing
                algorithm-agnostic parameters. User should not need to touch this.
```

**Comment by richardliaw** [comment]
BTW, I feel like this API exposes a bit too much. So far, the motivating use case for this has been just deadlines/time budget. But what else should we be exposing?

**Comment by Yard1** [comment]
IMO it's better to expose more than not enough. We don't know what exactly may be required by custom searchers/schedulers. This way we can future-proof the API, and considering that the Experiment object will not be saved in the Searcher/scheduler, I don't see any performance reason not to do it.

**Comment by richardliaw** [comment]
But this API allows us too much. I think we can expose fields/properties upon a case by case basis. 

 It would be very bad if we say in the future that we want to change some Experiment API or field like checkpoint freq, but searchers depend on it (when we should have exposed something else). Nowhere else is Experiment properly exposed iirc.

These requests are infrequent right now.

**Comment by Yard1** [comment]
@richardliaw That's a fair point. How about we pass the following fields from experiment, then:
- `stop`
- `time_budget_s`
- `resources_per_trial`
- `num_samples`
- `checkpoint_freq` (unsure)
- `checkpoint_at_end` (unsure)

I'd update the callback API PR then.

**Comment by richardliaw** [comment]
Great. Let’s avoid passing “checkpoint” params, as we don’t use that often
(we mainly use function api anyways, and those params don’t make sense for
function api)

**Comment by Yard1** [comment]
@xwjiang2010 Please see above. I'll make the change in my PR (https://github.com/ray-project/ray/pull/17794) and then it can be merged into this one to use the `get_public_attributes` method that will be added to `Experiment`.

**Comment by xwjiang2010** [comment]
Makes sense. I will patch yours and update.

File: python/ray/tune/suggest/ax.py [hunk 4]
```diff
@@ -230,8 +230,11 @@ def _setup_experiment(self):
             logger.warning("Detected sequential enforcement. Be sure to use "
                            "a ConcurrencyLimiter.")
 
-    def set_search_properties(self, metric: Optional[str], mode: Optional[str],
-                              config: Dict):
+    def set_search_properties(self,
+                              metric: Optional[str],
+                              mode: Optional[str],
+                              config: Dict,
+                              experiment: Optional[Dict] = None):
```
Comment (Yard1) [inline python/ray/tune/suggest/ax.py:237]: Maybe we can do
```suggestion
                              **spec):
```
instead

File: python/ray/tune/suggest/search.py [hunk 5]
```diff
@@ -37,12 +37,15 @@ def set_search_properties(self, metric: Optional[str], mode: Optional[str],
         Args:
             metric (str): Metric to optimize
             mode (str): One of ["min", "max"]. Direction to optimize.
-            config (dict): Tune config dict.
+            config (Dict): Tune config dict.
+            spec: Any kwargs for forward compatiblity.
```
Comment (Yard1) [inline python/ray/tune/suggest/search.py:41]:
```suggestion
            **spec: Any kwargs for forward compatiblity.
```

File: python/ray/tune/suggest/suggestion.py [hunk 6]
```diff
@@ -110,8 +111,11 @@ def __init__(self,
         else:
             raise ValueError("Mode most either be a list or string")
 
-    def set_search_properties(self, metric: Optional[str], mode: Optional[str],
-                              config: Dict) -> bool:
+    def set_search_properties(self,
+                              metric: Optional[str],
+                              mode: Optional[str],
+                              config: Dict,
+                              experiment: Optional[Dict] = None) -> bool:
```
Comment (Yard1) [inline python/ray/tune/suggest/suggestion.py:118]: Replace with **spec?

**Review by Yard1 — APPROVED** [review]
Looks good, just two nits!

File: python/ray/tune/suggest/bayesopt.py [hunk 7]
```diff
@@ -202,8 +202,11 @@ def _setup_optimizer(self):
         if self._analysis is not None:
             self.register_analysis(self._analysis)
 
-    def set_search_properties(self, metric: Optional[str], mode: Optional[str],
-                              config: Dict) -> bool:
+    def set_search_properties(self,
+                              metric: Optional[str],
+                              mode: Optional[str],
+                              config: Dict,
+                              experiment: Optional[Dict] = None) -> bool:
```
Comment (krfricke) [inline python/ray/tune/suggest/bayesopt.py:209]: For forwards compatibility, should we add `**spec` here, too? (and for the other searchers as well)

Comment (krfricke) [reply python/ray/tune/suggest/bayesopt.py:209]: Actually, shouldn't we use `**spec` here instead (and in all searchers except those actually using `experiment`)? Or alternatively use `experiment` everywhere

File: python/ray/tune/suggest/search.py [hunk 8]
```diff
@@ -37,12 +37,15 @@ def set_search_properties(self, metric: Optional[str], mode: Optional[str],
         Args:
             metric (str): Metric to optimize
             mode (str): One of ["min", "max"]. Direction to optimize.
-            config (dict): Tune config dict.
+            config (Dict): Tune config dict.
```
Comment (krfricke) [inline python/ray/tune/suggest/search.py:40]:
```suggestion
            config (Dict): Tune config dict.
            experiment (Experiment): Tune experiment object
```

File: python/ray/tune/suggest/search.py [hunk 9]
```diff
@@ -24,7 +24,7 @@ def metric(self):
         return self._metric
 
     def set_search_properties(self, metric: Optional[str], mode: Optional[str],
-                              config: Dict) -> bool:
+                              config: Dict, **spec) -> bool:
```
Comment (krfricke) [inline python/ray/tune/suggest/search.py:27]: IMO the base class should include all default arguments we actually pass to the child classes.

The child classes can then have `**spec` for forward compatibility

**Review by krfricke — COMMENTED** [review]
Generally looks good, let's just unify the argument definitions across searchers

File: python/ray/tune/suggest/search.py (hunk 9 above)
Comment (xwjiang2010) [reply python/ray/tune/suggest/search.py:27]: What's the downside of having **spec in base signature as well? If one is to implement a customized SearchAlgorithm, one should be able to figure out what is the right interface to implement against by looking at base SearchArgorithm class.

Comment (krfricke) [reply python/ray/tune/suggest/search.py:27]: Yes, that's why I would suggest to put `experiment` here, too (so people see that it is passed to searchers). We should also have `**spec` in that signature so when people inherit from this class they will keep the signature and hence keep forward compatibility

**Review by krfricke — APPROVED** [review]
Just one minor nit with the `experiment` argument in the base SearchAlgorithm interface, otherwise looks good!

File: python/ray/tune/suggest/search.py (hunk 9 above)
Comment (xwjiang2010) [reply python/ray/tune/suggest/search.py:27]: oh I see! Sorry was confused.
See Antoni's comment above (and Richard). I think we settled down on not passing experiment around and just use kwargs.

**Pull request merged**